frontend/javascripts/oxalis/model/bucket_data_handling/cuckoo_table.ts

```diff
@@ -0,0 +1,309 @@
+    ];
+    this.notifySeedListeners();
+  }
+
+  getTexture(): UpdatableTexture {
+    return this._texture;
+  }
+
+  subscribeToSeeds(fn: SeedSubscriberFn): () => void {
+    this.seedSubscribers.push(fn);
+    this.notifySeedListeners();
+
+    return () => {
+      this.seedSubscribers = this.seedSubscribers.filter((el) => el !== fn);
+    };
+  }
+
+  notifySeedListeners() {
+    this.seedSubscribers.forEach((fn) => fn(this.seeds));
+  }
+
+  getUniformValues() {
+    return {
+      CUCKOO_ENTRY_CAPACITY: this.entryCapacity,
+      CUCKOO_ELEMENTS_PER_ENTRY: ELEMENTS_PER_ENTRY,
+      CUCKOO_ELEMENTS_PER_TEXEL: TEXTURE_CHANNEL_COUNT,
+      CUCKOO_TWIDTH: this.textureWidth,
+    };
+  }
+
+  clearAll() {
+    this.table.fill(EMPTY_KEY);
+    this._texture.update(this.table, 0, 0, this.textureWidth, this.textureWidth);
+  }
+
+  set(pendingKey: number, pendingValue: Vector3, rehashAttempt: number = 0) {
+    if (pendingKey === EMPTY_KEY) {
+      throw new Error(`The key ${EMPTY_KEY} is not allowed for the CuckooTable.`);
+    }
+    let displacedEntry;
+    let currentAddress;
+    let iterationCounter = 0;
+
+    const ITERATION_THRESHOLD = 40;
+    const REHASH_THRESHOLD = 100;
+
+    if (rehashAttempt >= REHASH_THRESHOLD) {
+      throw new Error(
+        `Cannot rehash, since this is already the ${rehashAttempt}th attempt. Is the capacity exceeded?`,
+      );
+    }
+
+    let seedIndex = Math.floor(Math.random() * this.seeds.length);
+    while (iterationCounter++ < ITERATION_THRESHOLD) {
+      const seed = this.seeds[seedIndex];
+      currentAddress = this._hashKeyToAddress(seed, pendingKey);
+
+      // Swap pendingKey, pendingValue with what's contained in H1
+      displacedEntry = this.writeEntryAtAddress(
+        pendingKey,
+        pendingValue,
+        currentAddress,
+        rehashAttempt > 0,
+      );
+
+      if (this.canDisplacedEntryBeIgnored(displacedEntry[0], pendingKey)) {
+        return;
+      }
+
+      [pendingKey, pendingValue] = displacedEntry;
+
+      // Pick another random seed for the next swap
+      seedIndex =
+        (seedIndex + Math.floor(Math.random() * (this.seeds.length - 1)) + 1) % this.seeds.length;
+    }
+    this.rehash(rehashAttempt + 1);
+    this.set(pendingKey, pendingValue, rehashAttempt + 1);
+
+    // Since a rehash was performed, the incremental texture updates were
+    // skipped. Update the entire texture:
+    this._texture.update(this.table, 0, 0, this.textureWidth, this.textureWidth);
+  }
+
+  unset(key: number) {
+    for (const seed of this.seeds) {
+      const hashedAddress = this._hashKeyToAddress(seed, key);
+
+      const value = this.getValueAtAddress(key, hashedAddress);
+      if (value != null) {
+        this.writeEntryAtAddress(
+          EMPTY_KEY,
+          [EMPTY_KEY, EMPTY_KEY, EMPTY_KEY],
+          hashedAddress,
+          false,
+        );
+        return;
+      }
+    }
+  }
+
+  private rehash(rehashAttempt: number): void {
```

Comment:
"Newer hash functions" are achieved by generating three new seeds (this is done by `this.initializeTableArray();` which is called by the rehash method).

> As I understand it, each entry is simply reinserted and the only difference is that a different seed could be picked.

Since the actual seed instance properties are changed, the insertions are guaranteed to use different seeds (except for the rare case that the random seed generation picks the same seeds as before).

Does this make sense?

Reply:
Ah, sure, I overlooked the changing of the seeds. Makes perfect sense :+1: